# Advanced ML Models

# 1)- Import key modules

In [1]:
# support both Python 2 and Python 3 with minimal overhead.
from __future__ import absolute_import, division, print_function

# I am an engineer. I care only about error not warning. So, let's be maverick and ignore warnings.
import warnings
warnings.filterwarnings('ignore')

In [2]:
import re    # for regular expressions 
import nltk  # for text manipulation 
import string 
import numpy as np 
import pandas as pd 
import string 

#For Visuals
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
from matplotlib import rcParams
rcParams['figure.figsize'] = 11, 8
%config InlineBackend.figure_format = 'svg'
%matplotlib inline

In [3]:
#models and evaluation

from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from nltk.classify.scikitlearn import SklearnClassifier # notice its from ntlk not sklearn
from sklearn.svm import LinearSVC, SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
# Evaluation packages
from sklearn.metrics import accuracy_score
from sklearn import metrics
from sklearn.metrics import confusion_matrix

In [4]:
!pip install version_information

In [5]:
#pip install version_information
%reload_ext version_information
%version_information pandas,numpy, nltk, seaborn, matplotlib

Software versions
Python 3.6.9 64bit [GCC 8.4.0]
IPython 5.5.0
OS Linux 4.19.112+ x86_64 with Ubuntu 18.04 bionic
pandas 1.0.5
numpy 1.18.5
nltk 3.2.5
seaborn 0.10.1
matplotlib 3.2.2
Fri Sep 25 13:40:47 2020 UTC

In [6]:
# testing GPU on colab
import tensorflow as tf
tf.test.gpu_device_name()

''

# 2)- Loading Data

In [7]:
data=pd.read_csv('train_data_clean.csv')
#data=data.rename(columns={'Unnamed: 0':'random_columns'}) # a trick to tackle random index values
data.shape

(10000, 3)

In [8]:
data.head(2)

news  ...                                              clean
0  Top 5 Reasons Why 'Divergent' Star Kate Winsle...  ...  top 5 reason diverg star kate winslet deserv s...
1  Vessyl Bottle Tracks Your Drink And Its Health...  ...  vessyl bottl track drink health benefitsgadget...

[2 rows x 3 columns]

In [9]:
data.isnull().sum()

news        0
category    0
clean       0
dtype: int64

In [10]:
#loading test feature and label data saved from previous notebooks
feature_test=pd.read_csv('test_data.csv')
label_test=pd.read_csv('test_label.csv')

In [11]:
print(feature_test.shape)
print(label_test.shape)

(84484, 2)
(84484, 1)


# 3)- Vectorization

- bag of words
- tf-idf
- doc2vec
- word2vec

In [12]:
features=data['clean']
labels=data['category']
print(features.shape)
print(labels.shape)

(10000,)
(10000,)


### 3.1).Bag of Words

Bag-of-Words is a method to represent text into numerical features.

Let us understand this using a simple example. Suppose we have only 2 document

- D1: He is a lazy boy. She is also lazy.

- D2: Smith is a lazy person.

The list created would consist of all the unique tokens in the corpus C.

= [‘He’,’She’,’lazy’,’boy’,’Smith’,’person’]

Here, D=2, N=6



In [13]:
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
import gensim

bow_vectorizer = CountVectorizer(max_df=0.90, min_df=2, max_features=1000, stop_words='english')
# bag-of-words feature matrix
bow = bow_vectorizer.fit_transform(features)
bow.shape

(10000, 1000)

#### 3.1.a. Transform test data

In [14]:
feature_test.head(2)

Unnamed: 0                                              title
0      153245  iPhone 6 Release Date Pushed Back Due to Issue...
1      308611  Samsung Galaxy S4 vs Galaxy S3: Budget-Friendl...

In [15]:
bow_test = bow_vectorizer.transform(feature_test["title"])

In [16]:
print(bow.shape)
print(bow_test.shape)

(10000, 1000)
(84484, 1000)


### 3.2)-TF-IDF

This is another method which is based on the frequency method but it is different to the bag-of-words approach in the sense that it takes into account not just the occurrence of a word in a single document (or tweet) but in the entire corpus.

TF-IDF works by penalising the common words by assigning them lower weights while giving importance to words which are rare in the entire corpus but appear in good numbers in few documents.

Let’s have a look at the important terms related to TF-IDF:

- TF = (Number of times term t appears in a document)/(Number of terms in the document)

- IDF = log(N/n), where, N is the number of documents and n is the number of documents a term t has appeared in.

- TF-IDF = TF*IDF

In [17]:
tfidf_vectorizer = TfidfVectorizer(max_df=0.90, min_df=2, max_features=1000, stop_words='english')
# TF-IDF feature matrix
tfidf = tfidf_vectorizer.fit_transform(features)
tfidf.shape

(10000, 1000)

#### 3.2.a. Transform Test data

In [18]:
tfidf_test = tfidf_vectorizer.transform(feature_test["title"])

In [19]:
print(tfidf.shape)
print(tfidf_test.shape)

(10000, 1000)
(84484, 1000)


### 3.3)- Doc2Vec Embedding

In [20]:
from tqdm import tqdm 
tqdm.pandas(desc="progress-bar") 
from gensim.models.doc2vec import TaggedDocument

In [21]:
tokenized_text = data['clean'].apply(lambda x: x.split()) # tokenizing

In [22]:
def add_label(twt):
    output = []
    for i, s in zip(twt.index, twt):
        output.append(TaggedDocument(s, ["clean_" + str(i)]))
    return output
labeled_text = add_label(tokenized_text) # label all the news

##### 3.3.a.Train doc2vec model

In [23]:
model_d2v = gensim.models.Doc2Vec(dm=1,dm_mean=1,vector_size=200,window=5,negative=7,min_count=5,workers=3,alpha=0.1,seed=23)

In [24]:
model_d2v.build_vocab([i for i in tqdm(labeled_text)])

100%|██████████| 10000/10000 [00:00<00:00, 1565856.79it/s]


##### 3.3.b.Preparing doc2vec Feature Set

In [25]:
docvec_arrays = np.zeros((len(tokenized_text), 200))
for i in range(len(data)):
    docvec_arrays[i,:] = model_d2v.docvecs[i].reshape((1,200))

    
docvec_df = pd.DataFrame(docvec_arrays)
docvec_df.shape

(10000, 200)

In [26]:
docvec_df.head()

0         1         2    ...       197       198       199
0  0.002366  0.000339  0.001979  ... -0.001185  0.001691 -0.000495
1  0.002128  0.001104 -0.000378  ... -0.001520  0.001158  0.002369
2  0.000065 -0.000689  0.001343  ...  0.000314  0.001539 -0.001834
3  0.000201 -0.000361  0.001026  ... -0.001592 -0.001095 -0.002108
4 -0.001080  0.001557 -0.000543  ... -0.002416  0.002157  0.000595

[5 rows x 200 columns]

#### 3.3.3.transform test data

In [27]:
feature_test.head(2)

Unnamed: 0                                              title
0      153245  iPhone 6 Release Date Pushed Back Due to Issue...
1      308611  Samsung Galaxy S4 vs Galaxy S3: Budget-Friendl...

In [28]:
tokenized_text_test = feature_test['title'].apply(lambda x: x.split()) # tokenizing

In [29]:
labeled_text_test = add_label(tokenized_text_test) # label all the news

In [30]:
docvec_arrays_test = np.zeros((len(tokenized_text_test), 200))
for i in range(len(data)):
    docvec_arrays_test[i,:] = model_d2v.docvecs[i].reshape((1,200))

    
docvec_df_test = pd.DataFrame(docvec_arrays_test)


In [31]:
print(docvec_df.shape) # training
print(docvec_df_test.shape) # for testing

(10000, 200)
(84484, 200)


Notice I have only transformed test set and didn't train. So only my train set learns about vocab of corpus. My test model has only been transformed and learnt nothing.

### 3.4.Word2Vec Embedding

In [32]:
tokenized_text = data['clean'].apply(lambda x: x.split()) # tokenizing

model_w2v = gensim.models.Word2Vec(
            tokenized_text,
            size=200, # desired no. of features/independent variables
            window=5, # context window size
            min_count=2,
            sg = 1, # 1 for skip-gram model
            hs = 0,
            negative = 10, # for negative sampling i.e class with other types
            workers= 2, # no.of cores
            seed = 34) 

model_w2v.train(tokenized_text, total_examples= len(data['clean']), epochs=20)

(1329944, 1633380)

In [33]:
model_w2v.wv['nasdaq']

array([-0.02183259, -0.4366339 ,  0.06510326,  0.72783786,  0.24485947,
        0.33850783,  0.3416138 , -0.50301486,  0.29840896,  0.12920056,
        0.09041848, -0.50774693,  0.34069255, -0.43587837, -0.4792716 ,
        0.1502633 , -0.16197556,  0.3467329 , -0.26033154, -0.22837308,
       -0.2995486 ,  0.00264726,  0.32378572, -0.42868358,  0.03162241,
       -0.03673098, -0.34075603,  0.09162065, -0.15626894, -0.29774606,
        0.15991998, -0.05255641,  0.45747536, -0.31701863, -0.01571083,
       -0.13146812,  0.42375457, -0.31759384, -0.12677263,  0.25047785,
        0.44083124,  0.34026486,  0.02389309,  0.14316827,  0.26641372,
       -0.01980806, -0.14764051, -0.00484273,  0.03565803,  0.2957805 ,
       -0.00778876, -0.40688795, -0.72026014,  0.0653047 ,  0.13537495,
        0.16425957, -0.48244882, -0.5117245 ,  0.24898359, -0.30187702,
        0.29498556,  0.3112112 ,  0.10090353,  0.5491042 , -0.03915667,
        0.08010899, -0.2108901 , -0.051441  ,  0.5548677 ,  0.18

In [34]:
len(model_w2v.wv['nasdaq'])

200

In [35]:
type(model_w2v)

gensim.models.word2vec.Word2Vec

##### 3.4.1.Preparing Vectors for text data

In [36]:
def word_vector(tokens, size):
    vec = np.zeros(size).reshape((1, size))
    count = 0.
    for word in tokens:
        try:
            vec += model_w2v[word].reshape((1, size))
            count += 1.
        except KeyError: # handling the case where the token is not in vocabulary           
            continue
    if count != 0:
        vec /= count
    return vec

##### 3.4.2.Preparing word2vec feature set

In [37]:
wordvec_arrays = np.zeros((len(tokenized_text), 200)) 
for i in range(len(tokenized_text)):
    wordvec_arrays[i,:] = word_vector(tokenized_text[i], 200)
    wordvec_df = pd.DataFrame(wordvec_arrays)
wordvec_df.head()

0         1         2    ...       197       198       199
0  0.201549 -0.058880  0.098633  ...  0.029564 -0.295826 -0.001734
1  0.219597 -0.160597  0.166340  ...  0.150369  0.085569 -0.130088
2  0.068557 -0.153593  0.196505  ... -0.033059 -0.080726  0.033157
3  0.145885 -0.114159  0.145472  ...  0.103295 -0.073316 -0.070356
4  0.188864 -0.343754  0.201325  ...  0.360462 -0.056897 -0.303038

[5 rows x 200 columns]

In [38]:
wordvec_df.shape

(10000, 200)

#### 3.4.3.transform test data

In [39]:
feature_test.head(2)

Unnamed: 0                                              title
0      153245  iPhone 6 Release Date Pushed Back Due to Issue...
1      308611  Samsung Galaxy S4 vs Galaxy S3: Budget-Friendl...

In [40]:
tokenized_text_test = feature_test['title'].apply(lambda x: x.split()) # tokenizing

In [41]:
wordvec_arrays = np.zeros((len(tokenized_text_test), 200)) 
for i in range(len(tokenized_text_test)):
    wordvec_arrays[i,:] = word_vector(tokenized_text_test[i], 200)
    wordvec_df_test = pd.DataFrame(wordvec_arrays)
wordvec_df_test.head()

0         1         2    ...       197       198       199
0  0.208947 -0.076844  0.179379  ...  0.124041  0.060021  0.161705
1  0.289732 -0.248723  0.023323  ...  0.422009 -0.721807  0.372950
2  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000
3  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000
4  0.020771 -0.480825  0.128506  ...  0.145830 -0.117171 -0.132957

[5 rows x 200 columns]

In [42]:
print(wordvec_df.shape)
print(wordvec_df_test.shape)


(10000, 200)
(84484, 200)


It may look weird as we have less data for train and more to test. But, training is computing intense. So, this will help us. Plus eventually I will train my best performing model to whole data.

# 4)-Model Building

- Logistic Regression 
- Support Vector
- Random Forest
- XGBoost
- MLP

### 4.1.Logistic Regression Model

In [43]:
from sklearn.linear_model import LogisticRegression 
from sklearn.model_selection import train_test_split 
from sklearn.metrics import f1_score, accuracy_score, roc_curve,roc_auc_score,confusion_matrix, classification_report

##### 4.1.a. Logistic Regression using Bag-of-Words Features

In [44]:
X=bow
y=data['category']
print(X.shape)
print(y.shape)
print(bow_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [45]:
# splitting data into training and validation set
#xtrain_bow, xvalid_bow, ytrain, yvalid = train_test_split(bow, y,random_state=42,test_size=0.2)

In [46]:
lreg_bow = LogisticRegression(solver='liblinear')

# training the model
lreg_bow.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [47]:
# predicting on the validation set
prediction_bow = lreg_bow.predict_proba(bow_test)
prediction_bow[0]

array([0.16087434, 0.6072539 , 0.04823555, 0.1836362 ])

In [48]:
# prediction over classes

prediction_bow_class=lreg_bow.predict(bow_test)
prediction_bow_class[0]

'e'

In [49]:
accuracy_score(label_test, prediction_bow_class)

0.7044765872827992

In [50]:
from sklearn import metrics
print(metrics.classification_report(label_test, prediction_bow_class))

              precision    recall  f1-score   support

           b       0.73      0.65      0.69     23367
           e       0.65      0.92      0.76     30300
           m       0.75      0.47      0.58      9207
           t       0.80      0.56      0.66     21610

    accuracy                           0.70     84484
   macro avg       0.73      0.65      0.67     84484
weighted avg       0.72      0.70      0.70     84484



##### 4.1.b.Logistic Regression using TF-IDF Features

In [51]:
X=tfidf
y=data['category']
print(X.shape)
print(y.shape)
print(tfidf_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [52]:
lreg_tfidf = LogisticRegression(solver='liblinear')

# training the model
lreg_tfidf.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [53]:
# predicting on the validation set
prediction_tfidf = lreg_tfidf.predict_proba(tfidf_test)

In [54]:
prediction_tfidf_class=lreg_tfidf.predict(tfidf_test)

In [55]:
accuracy_score(label_test, prediction_tfidf_class)

0.707542256521945

In [56]:
print(metrics.classification_report(label_test, prediction_tfidf_class))

              precision    recall  f1-score   support

           b       0.73      0.66      0.69     23367
           e       0.66      0.92      0.77     30300
           m       0.78      0.46      0.58      9207
           t       0.79      0.57      0.66     21610

    accuracy                           0.71     84484
   macro avg       0.74      0.65      0.67     84484
weighted avg       0.72      0.71      0.70     84484



##### 4.1.c. Logistic Regression using Word2Vec Features

In [57]:
X=wordvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(wordvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [58]:
lreg_word2vec = LogisticRegression(solver='liblinear')
# training the model
lreg_word2vec.fit(X, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [59]:
# predicting on the validation set
prediction_word2vec = lreg_word2vec.predict_proba(wordvec_df_test)

In [60]:
prediction_word2vec_class=lreg_word2vec.predict(wordvec_df_test)

In [61]:
accuracy_score(label_test, prediction_word2vec_class)

0.4234884711898111

In [62]:
print(metrics.classification_report(label_test, prediction_word2vec_class))

              precision    recall  f1-score   support

           b       0.35      0.79      0.48     23367
           e       0.61      0.31      0.41     30300
           m       0.43      0.33      0.38      9207
           t       0.56      0.22      0.32     21610

    accuracy                           0.42     84484
   macro avg       0.49      0.41      0.40     84484
weighted avg       0.51      0.42      0.40     84484



##### 4.1.d. Logistic Regression using Doc2Vec Features

In [63]:
X=docvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(docvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [64]:
lreg_doc2vec = LogisticRegression(solver='liblinear')
# training the model
lreg_doc2vec.fit(docvec_df, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [65]:
# predicting on the validation set
prediction_doc2vec = lreg_doc2vec.predict_proba(docvec_df_test)

In [66]:
prediction_doc2vec_class=lreg_doc2vec.predict(docvec_df_test)

In [67]:
accuracy_score(label_test, prediction_doc2vec_class)

0.35864779129776053

In [68]:
print(metrics.classification_report(label_test, prediction_doc2vec_class))

              precision    recall  f1-score   support

           b       0.00      0.00      0.00     23367
           e       0.36      1.00      0.53     30300
           m       0.00      0.00      0.00      9207
           t       0.00      0.00      0.00     21610

    accuracy                           0.36     84484
   macro avg       0.09      0.25      0.13     84484
weighted avg       0.13      0.36      0.19     84484



**Summary:**

- bow=70%
- tfidf=70%
- word2vec=33%
- doc2vec=35%

### 4.2.Support Vector Machine (SVM)

In [69]:
X=bow
y=data['category']
print(X.shape)
print(y.shape)
print(bow_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [70]:
from sklearn import svm

##### SVM using Bag-of-Words Features

In [71]:
# we need to input arrays to our model.
bow_test = bow_vectorizer.transform(feature_test["title"]).toarray()

In [72]:
svc = svm.SVC(kernel='linear', C=1, probability=True).fit(X,y)

In [101]:
#prediction = svc.predict_proba(bow_test)

In [73]:
prediction_class = svc.predict(bow_test)

In [74]:
accuracy_score(label_test, prediction_class)

0.6848752426494957

In [75]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.71      0.64      0.67     23367
           e       0.63      0.92      0.75     30300
           m       0.72      0.45      0.56      9207
           t       0.81      0.50      0.62     21610

    accuracy                           0.68     84484
   macro avg       0.72      0.63      0.65     84484
weighted avg       0.71      0.68      0.67     84484



##### SVM using TF-IDF Features

In [76]:
X=tfidf
y=data['category']
print(X.shape)
print(y.shape)
print(tfidf_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [77]:
tfidf_test = tfidf_test.toarray()

In [78]:
svc = svm.SVC(kernel='linear',C=1, probability=True).fit(X,y)
#prediction = svc.predict_proba(xvalid_tfidf)
prediction_class = svc.predict(tfidf_test)
accuracy_score(label_test, prediction_class)

0.6975521992329908

In [79]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.72      0.65      0.68     23367
           e       0.65      0.91      0.76     30300
           m       0.74      0.46      0.57      9207
           t       0.79      0.55      0.65     21610

    accuracy                           0.70     84484
   macro avg       0.72      0.64      0.66     84484
weighted avg       0.71      0.70      0.69     84484



##### SVM using word2vec Features

In [80]:
X=wordvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(wordvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [81]:
svc = svm.SVC(kernel='linear', C=1, probability=True).fit(X,y)
#prediction = svc.predict_proba(wordvec_df_test)

In [82]:
prediction_class = svc.predict(wordvec_df_test)
accuracy_score(label_test, prediction_class)

0.42191420860754697

In [83]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.34      0.80      0.48     23367
           e       0.63      0.30      0.40     30300
           m       0.43      0.35      0.39      9207
           t       0.55      0.22      0.32     21610

    accuracy                           0.42     84484
   macro avg       0.49      0.42      0.40     84484
weighted avg       0.51      0.42      0.40     84484



##### SVM using doc2vec Features

In [84]:
X=docvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(docvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [85]:
svc = svm.SVC(kernel='linear', C=1, probability=True).fit(X,y)

In [86]:
#prediction = svc.predict_proba(xvalid_doc2vec)
prediction_class = svc.predict(docvec_df_test)
accuracy_score(label_test, prediction_class)

0.35864779129776053

In [87]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.00      0.00      0.00     23367
           e       0.36      1.00      0.53     30300
           m       0.00      0.00      0.00      9207
           t       0.00      0.00      0.00     21610

    accuracy                           0.36     84484
   macro avg       0.09      0.25      0.13     84484
weighted avg       0.13      0.36      0.19     84484



**Summary**


- bow = 69%
- tfidf= 68%
- word2vec= 42% 
- doc2vec= 41%

### 4.3.Random Forest

In [88]:
from sklearn.ensemble import RandomForestClassifier

##### RF with Bag-of-Words Features

In [89]:
X=bow
y=data['category']
print(X.shape)
print(y.shape)
print(bow_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [90]:
rf = RandomForestClassifier().fit(X, y)

In [91]:
#prediction = rf.predict_proba(xvalid_bow)
prediction_class = rf.predict(bow_test)
accuracy_score(label_test, prediction_class)

0.654088348089579

##### RF with TF-IDF Features

In [92]:
X=tfidf
y=data['category']
print(X.shape)
print(y.shape)
print(tfidf_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [93]:
rf = RandomForestClassifier().fit(X, y)

In [94]:
#prediction = rf.predict_proba(xvalid_tfidf)
prediction_class = rf.predict(tfidf_test)
accuracy_score(label_test, prediction_class)

0.6657591970077175

In [95]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.70      0.61      0.65     23367
           e       0.64      0.87      0.73     30300
           m       0.61      0.47      0.53      9207
           t       0.73      0.52      0.61     21610

    accuracy                           0.67     84484
   macro avg       0.67      0.62      0.63     84484
weighted avg       0.67      0.67      0.66     84484



##### RF with word2vec Features

In [96]:
X=wordvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(wordvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [97]:
rf = RandomForestClassifier().fit(X, y)
#prediction= rf.predict_proba(wordvec_df_test)
prediction_class = rf.predict(wordvec_df_test)
accuracy_score(label_test, prediction_class)

0.5087235452866815

In [98]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.61      0.38      0.47     23367
           e       0.47      0.87      0.61     30300
           m       0.55      0.27      0.36      9207
           t       0.55      0.25      0.34     21610

    accuracy                           0.51     84484
   macro avg       0.55      0.44      0.45     84484
weighted avg       0.54      0.51      0.48     84484



##### RF with doc2vec Feature

In [99]:
X=docvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(docvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [100]:
rf = RandomForestClassifier().fit(X, y)
prediction_class = rf.predict(docvec_df_test)
accuracy_score(label_test, prediction_class)

0.3488826286634156

In [101]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.27      0.03      0.06     23367
           e       0.36      0.92      0.52     30300
           m       0.11      0.01      0.02      9207
           t       0.26      0.03      0.06     21610

    accuracy                           0.35     84484
   macro avg       0.25      0.25      0.16     84484
weighted avg       0.28      0.35      0.22     84484



**Summary**

- bow = 57%
- tfidf = 56%
- word2vec = 57%
- doc2vec = 41%

# 4.4.XGBoost
Extreme Gradient Boosting (xgboost) is an advanced implementation of gradient boosting algorithm. It has both linear model solver and tree learning algorithms. Its ability to do parallel computation on a single machine makes it extremely fast. It also has additional features for doing cross validation and finding important variables. There are many parameters which need to be controlled to optimize the model.

Some key benefits of XGBoost are:

Regularization - helps in reducing overfitting
Parallel Processing - XGBoost implements parallel processing and is blazingly faster as compared to GBM.
Handling Missing Values - It has an in-built routine to handle missing values.
Built-in Cross-Validation - allows user to run a cross-validation at each iteration of the boosting process

**Notice there is no sklearn ready made model therefore; I needed to use XGBoost from its main librrary**

In [102]:
from xgboost import XGBClassifier

##### XGBoost using bag of words features

In [103]:
X=bow
y=data['category']
print(X.shape)
print(y.shape)
print(bow_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [104]:
xgb_model = XGBClassifier(max_depth=6, n_estimators=1000).fit(X, y)
prediction = xgb_model.predict_proba(bow_test)
prediction_class = xgb_model.predict(bow_test)
accuracy_score(label_test, prediction_class)

0.35864779129776053

In [105]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.00      0.00      0.00     23367
           e       0.36      1.00      0.53     30300
           m       0.00      0.00      0.00      9207
           t       0.00      0.00      0.00     21610

    accuracy                           0.36     84484
   macro avg       0.09      0.25      0.13     84484
weighted avg       0.13      0.36      0.19     84484



##### XGBoost using tfidf features

In [106]:
X=tfidf
y=data['category']
print(X.shape)
print(y.shape)
print(tfidf_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [107]:
xgb_model = XGBClassifier(max_depth=6, n_estimators=1000).fit(X, y)
prediction = xgb_model.predict_proba(tfidf_test)
prediction_class = xgb_model.predict(tfidf_test)
accuracy_score(label_test, prediction_class)

0.2557880782159936

In [108]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.00      0.00      0.00     23367
           e       0.00      0.00      0.00     30300
           m       0.00      0.00      0.00      9207
           t       0.26      1.00      0.41     21610

    accuracy                           0.26     84484
   macro avg       0.06      0.25      0.10     84484
weighted avg       0.07      0.26      0.10     84484



##### XGBoost using word2vecfeatures

In [109]:
X=wordvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(wordvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [110]:
xgb_model = XGBClassifier(max_depth=6, n_estimators=1000).fit(X, y)
prediction_class = xgb_model.predict(wordvec_df_test)
accuracy_score(label_test, prediction_class)

0.43125325505421147

In [111]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.61      0.39      0.48     23367
           e       0.60      0.33      0.43     30300
           m       0.51      0.32      0.40      9207
           t       0.30      0.66      0.41     21610

    accuracy                           0.43     84484
   macro avg       0.51      0.43      0.43     84484
weighted avg       0.52      0.43      0.44     84484



##### XGBoost using doc2vec features

In [112]:
X=docvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(docvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [113]:
xgb_model = XGBClassifier(max_depth=6, n_estimators=1000).fit(X, y)
prediction_class = xgb_model.predict(docvec_df_test)
accuracy_score(label_test, prediction_class)

0.3488826286634156

In [114]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.27      0.03      0.06     23367
           e       0.36      0.92      0.52     30300
           m       0.11      0.01      0.02      9207
           t       0.26      0.03      0.06     21610

    accuracy                           0.35     84484
   macro avg       0.25      0.25      0.16     84484
weighted avg       0.28      0.35      0.22     84484



**Summary**

- bow = 54%
- tfidf = 55%
- word2vec = 58%
- doc2vec = 37%

### 4.5.MLPClassifier

A multilayer perceptron (MLP) is a class of feedforward artificial neural network

In [115]:
from sklearn.neural_network import MLPClassifier

##### MLP using bag of words features

In [122]:
X=bow
y=data['category']
print(X.shape)
print(y.shape)
print(bow_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [123]:
mlp_model = MLPClassifier(random_state=1, max_iter=300,learning_rate_init=0.001).fit(X, y)
prediction = mlp_model.predict_proba(bow_test)
prediction_class = mlp_model.predict(bow_test)
accuracy_score(label_test, prediction_class)

0.6599356091094172

In [124]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.69      0.59      0.64     23367
           e       0.68      0.81      0.74     30300
           m       0.49      0.53      0.51      9207
           t       0.68      0.58      0.63     21610

    accuracy                           0.66     84484
   macro avg       0.63      0.63      0.63     84484
weighted avg       0.66      0.66      0.66     84484



##### MLP using tfidf features

In [119]:
X=tfidf
y=data['category']
print(X.shape)
print(y.shape)
print(tfidf_test.shape)
print(label_test.shape)

(10000, 1000)
(10000,)
(84484, 1000)
(84484, 1)


In [120]:
mlp_model = MLPClassifier(random_state=1, max_iter=300,learning_rate_init=0.001).fit(X, y)
prediction_class = mlp_model.predict(tfidf_test)
accuracy_score(label_test, prediction_class)

0.6671085649353724

In [121]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.69      0.61      0.65     23367
           e       0.67      0.83      0.74     30300
           m       0.57      0.49      0.53      9207
           t       0.68      0.57      0.62     21610

    accuracy                           0.67     84484
   macro avg       0.65      0.63      0.63     84484
weighted avg       0.67      0.67      0.66     84484



##### MLP using word2vecfeatures

In [125]:
X=wordvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(wordvec_df_test.shape)
print(label_test.shape)

(10000, 200)
(10000,)
(84484, 200)
(84484, 1)


In [126]:
mlp_model = MLPClassifier(random_state=1, max_iter=300,learning_rate_init=0.001).fit(X, y)
prediction_class = mlp_model.predict(wordvec_df_test)
accuracy_score(label_test, prediction_class)

0.4867548884995976

In [127]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.53      0.39      0.45     23367
           e       0.47      0.81      0.59     30300
           m       0.44      0.32      0.37      9207
           t       0.55      0.21      0.30     21610

    accuracy                           0.49     84484
   macro avg       0.50      0.43      0.43     84484
weighted avg       0.50      0.49      0.45     84484



##### MLP using doc2vec features

In [ ]:
X=docvec_df
y=data['category']
print(X.shape)
print(y.shape)
print(docvec_df_test.shape)
print(label_test.shape)

In [ ]:
mlp_model = MLPClassifier(random_state=1, max_iter=300,learning_rate_init=0.001).fit(X, y)
prediction_class = mlp_model.predict(docvec_df_test)
accuracy_score(label_test, prediction_class)

0.4117647058823529

In [128]:
print(metrics.classification_report(label_test, prediction_class))

              precision    recall  f1-score   support

           b       0.53      0.39      0.45     23367
           e       0.47      0.81      0.59     30300
           m       0.44      0.32      0.37      9207
           t       0.55      0.21      0.30     21610

    accuracy                           0.49     84484
   macro avg       0.50      0.43      0.43     84484
weighted avg       0.50      0.49      0.45     84484



**Summary**

- bow = 65%
- tfidf = 66%
- word2vec = 48%
- doc2vec = 41%

# END OF NOTEBOOK CODE